src/eval/datetime.rs

```diff
@@ -0,0 +1,110 @@
+use ecow::{eco_vec, EcoVec};
+use std::fmt;
+use std::fmt::{Debug, Formatter};
+use time::error::{Format, InvalidFormatDescription};
+use time::format_description;
+use typst_macros::cast_from_value;
+
+#[derive(Clone, Copy, PartialEq, Hash)]
+pub enum Datetime {
+    Date(time::Date),
+    Datetime(time::PrimitiveDateTime),
+    Time(time::Time),
+}
+
+impl Datetime {
+    pub fn display(&self, pattern: Option<String>) -> Result<String, String> {
+        let pattern = pattern.unwrap_or(match self {
+            Datetime::Date(_) => String::from("[year]-[month]-[day]"),
+            Datetime::Time(_) => String::from("[hour]:[minute]:[second]"),
+            Datetime::Datetime(_) => {
+                String::from("[year]-[month]-[day] [hour]:[minute]:[second]")
+            }
+        });
+
+        let format =
+            format_description::parse(pattern.as_str()).map_err(|e| match e {
+                InvalidFormatDescription::UnclosedOpeningBracket { .. } => {
+                    "found unclosed bracket".to_string()
+                }
+                InvalidFormatDescription::InvalidComponentName { name, .. } => {
+                    format!("{} is not a valid component", name)
+                }
+                InvalidFormatDescription::InvalidModifier { value, .. } => {
+                    format!("modifier {} is invalid", value)
+                }
+                InvalidFormatDescription::Expected { what, .. } => {
+                    format!("expected {}", what)
+                }
+                InvalidFormatDescription::MissingComponentName { .. } => {
+                    "a component name is missing".to_string()
+                }
+                InvalidFormatDescription::MissingRequiredModifier { name, .. } => {
+                    format!("missing required modifier {}", name)
+                }
+                InvalidFormatDescription::NotSupported { context, what, .. } => {
+                    format!("{} is not supported in {}", what, context)
+                }
+                _ => "unable to parse datetime format".to_string(),
+            })?;
+
+        let formatted_result = match self {
+            Datetime::Date(date) => date.format(&format),
+            Datetime::Time(time) => time.format(&format),
+            Datetime::Datetime(datetime) => datetime.format(&format),
+        };
+
+        let unwrapped_result = formatted_result.map_err(|e| match e {
+            Format::InvalidComponent(name) => format!("found invalid component {}", name),
+            _ => "unable to format datetime in the requested format".to_string(),
```

Comment:
The thing is, if an error occurs there there probably isn't muth the user can do anyway. This is the source code of the error in the `time` crate:

https://github.com/time-rs/time/blob/e83d01fa97bebc544473c5b99d80be98956fcb38/time/src/error/format.rs#L9-L22

I can't handle the `InsufficientTimeInformation` one separately, because Rust tells me it's private (maybe due to the `non_exhaustive`). So this only leaves `Std::Io`, I admittedly haven't looked into how exactly this one could occur, but this probably is a pretty low-level error which doesn't make sense to be passed on to the user. So I think it would be better to just return a generic error message and let users create an issue for it so that it can be investigated.